File: apps/api/app/services/exports.py
from datetime import UTC, datetime
from io import BytesIO

from openpyxl import Workbook
from openpyxl.styles import Alignment, Font, PatternFill
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.db.base import AuditEvent, QualityIssue, User
from app.domain.decimal_utils import decimal_to_str
from app.domain.simulation import simulate
from app.domain.types import Driver
from app.schemas.api import ExportRequest
from app.services.analysis import analyze
from app.services.persistence import latest_batch

INK = "1D1D1B"
PETROL = "224957"
TEAL = "379B8C"
YELLOW = "E5AD29"
ORANGE = "EA782E"
RED = "E43023"
IVORY = "F7F5EF"


def create_export(db: Session, request: ExportRequest, actor: User) -> BytesIO:
    analysis = analyze(db, request.query_date, request.price_list_code)
    batch = latest_batch(db)
    workbook = Workbook()
    results = workbook.active
    results.title = "Resultados"
    observations = workbook.create_sheet("Observaciones")
    metadata = workbook.create_sheet("Metadatos")

    result_headers = [
        "Fecha consulta", "Lista", "Producto", "Descripción", "Sucursal",
        "Costo vigente", "Vigencia costo", "Precio original", "Vigencia precio",
        "Margen ideal %", "Conductor", "Precio simulado", "Ganancia $", "Ganancia %",
        "Diferencia $", "Diferencia p.p.", "Termómetro", "Estado", "Advertencias",
    ]
    results.append(result_headers)
    for row in analysis.rows:
        simulation_input = request.simulations.get(row.product_code)
        calculated = None
        if simulation_input is not None and not row.simulation_blocked:
            calculated = simulate(
                cost=simulation_input.cost,
                driver=Driver(simulation_input.driver),
                driver_value=simulation_input.driver_value,
                ideal_percent=simulation_input.ideal_percent,
                source_inactive=simulation_input.source_inactive,
                source_unknown=simulation_input.source_unknown,
            )
        results.append([
            request.query_date,
            row.price_list_name,
            row.product_code,
            row.description,
            row.branch_code,
            row.cost.value,
            row.cost.valid_from,
            row.price.value,
            row.price.valid_from,
            row.ideal_percent,
            simulation_input.driver if simulation_input else None,
            decimal_to_str(calculated.price) if calculated else None,
            decimal_to_str(calculated.gain_amount) if calculated else row.actual_gain_amount,
            decimal_to_str(calculated.gain_percent) if calculated else row.actual_gain_percent,
            decimal_to_str(calculated.gap_amount) if calculated else None,
            decimal_to_str(calculated.gap_percentage_points) if calculated else None,
            calculated.thermometer.value if calculated else "neutral",
            row.data_status,
            ", ".join(row.warnings),
        ])

    observations.append(["Tipo", "Severidad", "Hoja", "Clave", "Explicación", "Filas", "Valores"])
    if batch:
        for issue in db.scalars(select(QualityIssue).where(QualityIssue.batch_id == batch.id)).all():
            observations.append([
                issue.issue_type, issue.severity, issue.sheet_name, issue.business_key,
                issue.explanation, ", ".join(map(str, issue.source_rows)), ", ".join(str(v) for v in issue.values),
            ])

    metadata.append(["Campo", "Valor"])
    metadata.append(["Lote", batch.id if batch else ""])
    metadata.append(["Archivo", batch.filename if batch else ""])
    metadata.append(["Fecha de consulta", request.query_date])
    metadata.append(["Lista", analysis.price_list.description])
    metadata.append(["Exportado por", actor.email])
    metadata.append(["Fecha de exportación", datetime.now(UTC).isoformat()])
    metadata.append(["Uso", "Archivo analítico; no apto para carga automática en TOTVS."])

    for sheet in workbook.worksheets:
        sheet.freeze_panes = "A2"
        sheet.sheet_view.showGridLines = False
        for cell in sheet[1]:
            cell.fill = PatternFill("solid", fgColor=PETROL)
            cell.font = Font(color="FFFFFF", bold=True)
            cell.alignment = Alignment(vertical="center")
        for column in sheet.columns:
            letter = column[0].column_letter
            width = min(max(len(str(cell.value or "")) for cell in column) + 2, 42)
            sheet.column_dimensions[letter].width = max(width, 12)

    db.add(AuditEvent(actor_id=actor.id, action="export.created", entity_type="import_batch", entity_id=batch.id if batch else None, details={"date": str(request.query_date), "price_list": request.price_list_code, "rows": len(analysis.rows)}))
    db.commit()
    buffer = BytesIO()
    workbook.save(buffer)
    buffer.seek(0)
    return buffer

File: apps/web/lib/engine/exports.ts
// Exportación analítica: genera un Excel de tres hojas (Resultados,
// Observaciones, Metadatos). Recalcula las simulaciones visibles en el servidor
// (autoridad). Archivo analítico; NO apto para carga automática en TOTVS.
import ExcelJS from "exceljs";
import { Decimal, toStr } from "./decimal";
import { simulate } from "./simulation";
import type { Driver } from "./types";
import type { AnalysisResponse } from "./analysis";
import type { ParsedIssue } from "./importer";

const PETROL = "FF224957";

export interface SimulationExportInput {
  cost: string | null;
  idealPercent: string | null;
  driver: Driver;
  driverValue: string;
  sourceInactive?: boolean;
  sourceUnknown?: boolean;
}

export interface ExportInput {
  queryDate: string;
  analysis: AnalysisResponse;
  issues: ParsedIssue[];
  simulations?: Record<string, SimulationExportInput>;
  metadata: { batchId: string | null; filename: string | null; exportedBy: string; exportedAt: string };
}

const dec = (s: string | null): Decimal | null => (s === null || s.trim() === "" ? null : new Decimal(s));

export async function buildExportWorkbook(input: ExportInput): Promise<Buffer> {
  const { analysis, queryDate } = input;
  const wb = new ExcelJS.Workbook();
  wb.creator = "Brugali · Costos y precios";
  wb.created = new Date(input.metadata.exportedAt);

  // ---- Resultados ----
  const results = wb.addWorksheet("Resultados");
  results.addRow([
    "Fecha consulta", "Lista", "Producto", "Descripción", "Sucursal",
    "Costo vigente", "Vigencia costo", "Precio original", "Vigencia precio",
    "Margen ideal %", "Conductor", "Precio simulado", "Ganancia $", "Ganancia %",
    "Diferencia $", "Diferencia p.p.", "Termómetro", "Estado", "Advertencias",
  ]);

  for (const row of analysis.rows) {
    const simInput = input.simulations?.[row.productCode];
    let calc: ReturnType<typeof simulate> | null = null;
    if (simInput && !row.simulationBlocked) {
      calc = simulate({
        cost: dec(simInput.cost),
        driver: simInput.driver,
        driverValue: new Decimal(simInput.driverValue || "0"),
        idealPercent: dec(simInput.idealPercent),
        sourceInactive: simInput.sourceInactive,
        sourceUnknown: simInput.sourceUnknown,
      });
    }
    results.addRow([
      queryDate,
      row.priceListName,
      row.productCode,
      row.description,
      row.branchCode,
      row.cost.value,
      row.cost.validFrom,
      row.price.value,
      row.price.validFrom,
      row.idealPercent,
      simInput ? simInput.driver : null,
      calc ? toStr(calc.price) : null,
      calc ? toStr(calc.gainAmount) : row.actualGainAmount,
      calc ? toStr(calc.gainPercent) : row.actualGainPercent,
      calc ? toStr(calc.gapAmount) : null,
      calc ? toStr(calc.gapPercentagePoints) : null,
      calc ? calc.thermometer : "neutral",
      row.dataStatus,
      row.warnings.join(", "),
    ]);
  }

  // ---- Observaciones ----
  const observations = wb.addWorksheet("Observaciones");
  observations.addRow(["Tipo", "Severidad", "Hoja", "Clave", "Explicación", "Filas", "Valores"]);
  for (const issue of input.issues) {
    observations.addRow([
      issue.issueType,
      issue.severity,
      issue.sheetName,
      issue.businessKey,
      issue.explanation,
      issue.sourceRows.join(", "),
      issue.values.map((v) => (v === null ? "" : v)).join(", "),
    ]);
  }

  // ---- Metadatos ----
  const metadata = wb.addWorksheet("Metadatos");
  metadata.addRow(["Campo", "Valor"]);
  metadata.addRow(["Lote", input.metadata.batchId ?? ""]);
  metadata.addRow(["Archivo", input.metadata.filename ?? ""]);
  metadata.addRow(["Fecha de consulta", queryDate]);
  metadata.addRow(["Lista", analysis.priceList.description]);
  metadata.addRow(["Exportado por", input.metadata.exportedBy]);
  metadata.addRow(["Fecha de exportación", input.metadata.exportedAt]);
  metadata.addRow(["Uso", "Archivo analítico; no apto para carga automática en TOTVS."]);

  for (const sheet of wb.worksheets) {
    sheet.views = [{ state: "frozen", ySplit: 1 }];
    const header = sheet.getRow(1);
    header.eachCell((cell) => {
      cell.fill = { type: "pattern", pattern: "solid", fgColor: { argb: PETROL } };
      cell.font = { color: { argb: "FFFFFFFF" }, bold: true };
      cell.alignment = { vertical: "middle" };
    });
    sheet.columns.forEach((col) => {
      let width = 12;
      col.eachCell?.({ includeEmpty: false }, (cell) => {
        width = Math.max(width, String(cell.value ?? "").length + 2);
      });
      col.width = Math.min(width, 42);
    });
  }

  return Buffer.from((await wb.xlsx.writeBuffer()) as unknown as ArrayBuffer);
}
